### ISSUE TITLE
Report problems with 1.6 updates here.

### ISSUE THREAD

**Issue by kdschlosser** [OpenZWave/python-openzwave#181]
This issue is for anyone that decides they want to test my fork located here for any issues. I want to keep the commit history clean so i will sort out any problems and then submit a PR.

You can clone the repo from here https://github.com/kdschlosser/python-openzwave/tree/1.6
make sure you clone the 1.6 branch.

You will need to use the --flavor=dev argument so it will download the latest code from the OpenZWave

**Comment by petergebruers** [comment]
Thank you for sharing this. I am trying to make this work with hass in a venv + my fork of ozw. pip install and cythonizing is OK. That's good news. I am on macos, fwiw. When hass started, it barked:

```
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/command.py", line 32, in <module>
    from .scene import ZWaveSceneList
```

I commented out that line in command.py, I think it is a leftover from 1.4, isn't it?

Second issue, if I erase my hass config, then add Z-Wave, I get a recursion depth error when I start hass:

```
 Traceback (most recent call last):
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/homeassistant/config_entries.py", line 309, in async_setup
    hass, self)
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/homeassistant/components/zwave/__init__.py", line 290, in async_setup_entry
    network = hass.data[DATA_NETWORK] = ZWaveNetwork(options, autostart=False)
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/singleton.py", line 49, in __call__
    super(InstanceSingleton, cls).__call__(*args, **kwargs)
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/network.py", line 319, in __init__
    self._manager = ZWaveManager()
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/manager.py", line 67, in __init__
    self.__manager = libopenzwave.PyManager()
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/manager.py", line 87, in __setattr__
    setattr(self.__manager, key, value)
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/manager.py", line 73, in __getattr__
    if hasattr(self.__manager, item):
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/manager.py", line 73, in __getattr__
    if hasattr(self.__manager, item):
  File "/Users/peter/hass_venv/lib/python3.7/site-packages/openzwave/manager.py", line 73, in __getattr__
    if hasattr(self.__manager, item):
  [Previous line repeated 487 more times]
RecursionError: maximum recursion depth exceeded
```

My system is a mess because of all the development going on, so there's always the possibility that it is not a bug but my environment... Unfortunately I cannot work on this right now, I have a few other tasks to complete.

**Comment by kdschlosser** [comment]
no these are issues with the library.

I am making a new version of it. that has less of the original code disturbed. I had to make some changes and then revert things back and try something different and then revert back. after so many times of doing that it ends up being a mess. And there are things I would forget to revert I am sure.

so the scene list thing is from a test that I did. but is not needed. I am guessing that I have a remnant of it still kicking about.

the manager things is because I created a wrapper class that then wraps all call to manager to catch the exceptions. I must have an issue in there that is causing some kind of a recursion problem. I will get it sorted.

**Comment by kdschlosser** [comment]
and I see where the recursion problem is already. LOL.

**Comment by kdschlosser** [comment]
OK I have cleaned up the code. broken apart the commits better. change the file structure by splitting off any of the command class specific properties/methods into their own files. the command.py file was getting a little beefy for my liking. making it hard to find things.

I did expose the SendRawData. it is kind of hidden. it's not something that was added directly to the public API. to access it you would need to use `network.manager.sendRawData()`

Everything should have docstrings and those docstrings have a stabndard of formatting done to them.

I added my special assortment of utilities that i have made over the last year. i removed the old deprecated decorator as this was only able to be used on methods and functions. The one I created works on classes/methods/functions/nested functions/nested classes/properties - get and set, individually or together/variables with any data type. the ability to set a custom message is also available.

in that bunch of utilities is also a series of decorators for debugging the data path through a program. It works on methods, functions and properties (get, set and delete). it logs any data passed as parameters. the keywords to those parameters. the file name and line number of where the call was made from. the file name and the line number of the wrapped function. you also have the option of logging the returns data as well. there is also a timer decorator to make it easier to aid in locating inefficient code. I have not implemented any of these extra debugging decorators as of yet. It will be something that will get used specifically the timing decorator. need to see if we can boost performance of this program. 

All of the updates that are needed for 1.6 have been added. I have not yet tested it. if someone is wanting to give it a try use the "new-1.6" branch in my fork. This branch contains all of the code updates that are in the python-openzwave 0.5.x branch. so it has a whole new windows build system. along with auto detection of the serial port the ZStick is using. When passing the paths to the config file locations the path can now be relative it does not have to be absolute. the expansion of variables now takes place as well. so you NIX folks can use ~ and windows folks can use the %VAR%. NIX can also use $VAR, Windows users no longer need to specify the virtual namespace of the com port. using just COM15 will now work. NIX and Windows can pass None as the port and this will start the detection process. currently this only works for the sillabs chips. If you have something different please pass on over the VID and PID for the stick.

You no longer need to worry about encoding and decoding strings, bytes, unicode before passing them to any of the components on python-openzwave. this is now handled internally. I am thinking there is going to be quite a few developers that are going to be happy because of this.

Most of the get_ and set_ methods have been depreciated. they are now properties. I was able to remove the value_id needing to be passed to them with the code change for accessing the individual values by index. So developers will no longer need to keep track of value_id to what it does. I added a whole slew of additional properties/methods for setting specific values here is a list of them.

COMMAND_CLASS_SIMPLE_AV_CONTROL
(meth) av_command_send
(prop get) av_commands

COMMAND_CLASS_CENTRAL_SCENE
(prop get) scenes

COMMAND_CLASS_DOOR_LOCK
(prop get) door_locks
(prop get/set) door_lock
(prop get) user_codes
(prop get) doorlock_logs

COMMAND_CLASS_NOTIFICATION (COMMAND_CLASS_ALARM)
(prop get) alarm_state
(prop get) alarm_level
(prop get) alarm_type
(prop get) alarm_parameters

COMMAND_CLASS_PROTECTION
(prop get/set) protection
(prop get) protections
(prop get) protection_items

COMMAND_CLASS_SENSOR_BINARY, COMMAND_CLASS_SENSOR_MULTILEVEL, COMMAND_CLASS_METER
(prop get) sensors
(prop get) sensor_value

COMMAND_CLASS_SOUND_SWITCH
(prop get/set) volume
(prop get/set) tone
(prop get) tone_items

COMMAND_CLASS_SWITCH_ALL
(prop get) switches_all
(prop get/set) switch_all_state
(prop get) switch_all_items

COMMAND_CLASS_SWITCH_BINARY
(prop get) switches
(prop get/set) switch_state

COMMAND_CLASS_SWITCH_MULTILEVEL
(prop get) dimmers
(prop get/set) dimmer_level

COMMAND_CLASS_COLOR
(prop get) rgb_bulbs
(prop get/set) rgbw

COMMAND_CLASS_THERMOSTAT_MODE
(prop get/set) thermostat_operating_mode

COMMAND_CLASS_THERMOSTAT_OPERATING_STATE
(prop get) thermostat_operating_state

COMMAND_CLASS_THERMOSTAT_SETPOINT
(prop get/set) thermostat_heating
(prop get/set) thermostat_economy_heating
(prop get/set) thermostat_away_heating
(prop get/set) thermostat_cooling
(prop get/set) thermostat_economy_cooling
(prop get/set) thermostat_away_cooling

COMMAND_CLASS_THERMOSTAT_FAN_MODE
(prop get/set) thermostat_fan_mode

COMMAND_CLASS_THERMOSTAT_FAN_STATE
(prop get) thermostat_fan_state

these nest 2 cover all of the THERMOSTAT command classes 
(prop get) thermostats
(prop get) thermostat_state


all of the original methods are still in place and have not been removed. the API has not been broken. they are now deprecated.

All class instances created in python-openzwave are instance singletons. if a class instance exists that is what will be returned, a new instance will not be created. This will ensure any data that is only held within python-openzwave will be the same no matter where you access it from in your code. so holding references to instances is OK to do. I have not done this yet but I will be soon is adding in a mechanism into nodes and values where it will become non operational if removed from the network. so setting anything in that instance will simply do nothing.

I did create a wrapper for all calls made into OZW. this wrapper has the sole purpose of exception catching any calls made. a traceback will be printed out but it will not cause the program to stop running.

I added an index mapping to enumeration name tho I did not make this portion of the code public. I have not decided on whether or not it should be a private thing or not. I am looking for some input from people on which way to go with that portion of the code. It can still be accessed even tho it is "private" I did not do any mangling of the name. 

The user code portion of the door locks has been completely redesigned. when you use the property "user_codes" it is going to return an instance of a class that is made to work like a python list. so codes can be gotten and set by using the standard list index `user_codes[1]` there is also append extend and remove. I did add the ability to change the label associated with a user code so if wanted you can supply a user friendly name to a code. and you can access a code in the standard dictionary `user_codes['name']` fashion. the code can also be set using the same fashion so long as the label has been set. you can also use `len(user_codes)` and `'name' in user_codes` as well as `user_codes += ['code1', 'code2']`

**Comment by kdschlosser** [comment]
I still have a few small glitches to work out. in the program but nothing major.

I did make a modification to the setup program. There is no longer the need to have the user install any modules. Everything is done internally in the setup program. It is nice to have it this way. It makes it easier for anyone adding the library to another project. they can now just add it as a requirement.

**Comment by kdschlosser** [comment]
I rewrote the config file installation portion of the setup program. the old one would blindly delete the config file directory if an upgrade or a reinstall was being done. I did not personally care for this because of the chance of having someone's files getting deleted. So now if there is a previous installation it will backup the config files and then place them back into place. then it performs a comparison of the 2 files. the first comparison is for equality. if they are the same the program continues. if they are not the same. then it looks for a revision number in both of the file. if both files have a revision it goes with the one that has the larger number. if either of the files does not have a revision number it will then go with the larger of the 2 files.

This is the best mechanism I can think of to protect a users config files in the event they made any changes. There is no sure fire way of knowing which one is the best to use. but this is going to cover most scenarios I think.
